File: frontend/lib/inventory.ts
import { api, encodeId } from "@/lib/http";

// ---- Types ----
export type Warehouse = {
  id: string;
  name: string;
  parent_warehouse: string | null;
  is_group: boolean;
  disabled: boolean;
};
export type WarehouseInput = {
  name: string;
  parent_warehouse?: string | null;
  is_group?: boolean;
  disabled?: boolean;
};

export type Batch = {
  id: string;
  batch_id: string;
  item_code: string;
  expiry_date: string | null;
  manufacturing_date: string | null;
  qty: number | null;
};
export type BatchInput = {
  batch_id: string;
  item_code: string;
  expiry_date?: string | null;
  manufacturing_date?: string | null;
};

export type StockLevel = {
  item_code: string;
  warehouse: string;
  actual_qty: number;
  reserved_qty: number | null;
  projected_qty: number | null;
};

export type MovementLine = {
  item_code: string;
  qty: number;
  batch_no?: string | null;
  rate?: number | null;
};
export type MovementInput = { warehouse: string; items: MovementLine[] };

export type StockEntry = {
  id: string;
  stock_entry_type: string;
  items: {
    item_code: string;
    qty: number;
    warehouse: string | null;
    batch_no: string | null;
  }[];
};

// ---- Warehouses ----
export const listWarehouses = (token: string) =>
  api.get<Warehouse[]>(token, "/inventory/warehouses", { limit: 200 });

export const createWarehouse = (token: string, input: WarehouseInput) =>
  api.post<Warehouse>(token, "/inventory/warehouses", input);

export const updateWarehouse = (token: string, id: string, input: WarehouseInput) =>
  api.put<Warehouse>(token, `/inventory/warehouses/${encodeId(id)}`, input);

export const deleteWarehouse = (token: string, id: string) =>
  api.del(token, `/inventory/warehouses/${encodeId(id)}`);

// ---- Batches ----
export const listBatches = (token: string) =>
  api.get<Batch[]>(token, "/inventory/batches", { limit: 200 });

export const createBatch = (token: string, input: BatchInput) =>
  api.post<Batch>(token, "/inventory/batches", input);

// ---- Stock ----
export const listStock = (
  token: string,
  params: { item_code?: string; warehouse?: string } = {}
) => api.get<StockLevel[]>(token, "/inventory/stock", { ...params, limit: 500 });

// ---- Goods movements ----
export const receiveGoods = (token: string, input: MovementInput) =>
  api.post<StockEntry>(token, "/inventory/receive", input);

export const issueGoods = (token: string, input: MovementInput) =>
  api.post<StockEntry>(token, "/inventory/issue", input);

File: backend/schemas/inventory.py
"""Inventory schemas (Pydantic V2).

Maps onto ERPNext DocTypes:
  - Warehouse    -> warehouses
  - Batch        -> batch numbers + expiry dates
  - Bin          -> product stock levels (read-only, computed by ERPNext)
  - Stock Entry  -> goods received (Material Receipt) / goods issued (Material Issue)
Field mappings live in the corresponding repositories.
"""

from datetime import date

from pydantic import BaseModel, Field

# ---------------------------------------------------------------- Warehouse


class WarehouseBase(BaseModel):
    name: str = Field(min_length=1, max_length=140)  # -> warehouse_name
    parent_warehouse: str | None = None
    is_group: bool = False
    disabled: bool = False


class WarehouseCreate(WarehouseBase):
    pass


class WarehouseUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=140)
    parent_warehouse: str | None = None
    is_group: bool | None = None
    disabled: bool | None = None


class WarehouseRead(WarehouseBase):
    id: str  # ERPNext Warehouse name


# -------------------------------------------------------------------- Batch


class BatchBase(BaseModel):
    batch_id: str = Field(min_length=1, max_length=140)
    item_code: str = Field(min_length=1, max_length=140)  # -> item
    expiry_date: date | None = None
    manufacturing_date: date | None = None


class BatchCreate(BatchBase):
    pass


class BatchRead(BatchBase):
    id: str
    qty: float | None = None  # batch_qty (computed)


# --------------------------------------------------------------- Stock level


class StockLevel(BaseModel):
    item_code: str
    warehouse: str
    actual_qty: float = 0
    reserved_qty: float | None = None
    projected_qty: float | None = None


# ------------------------------------------------------- Goods movements


class MovementLine(BaseModel):
    item_code: str = Field(min_length=1, max_length=140)
    qty: float = Field(gt=0)
    batch_no: str | None = None
    rate: float | None = Field(default=None, ge=0)  # basic_rate (receipts)


class GoodsReceiptRequest(BaseModel):
    """Goods Received -> ERPNext Stock Entry (Material Receipt)."""

    warehouse: str = Field(min_length=1)  # target warehouse
    items: list[MovementLine] = Field(min_length=1)


class GoodsIssueRequest(BaseModel):
    """Goods Issued -> ERPNext Stock Entry (Material Issue)."""

    warehouse: str = Field(min_length=1)  # source warehouse
    items: list[MovementLine] = Field(min_length=1)


class StockEntryLine(BaseModel):
    item_code: str
    qty: float
    warehouse: str | None = None
    batch_no: str | None = None


class StockEntryRead(BaseModel):
    id: str
    stock_entry_type: str
    items: list[StockEntryLine]
